### PR TITLE
Fix mapped type instantiation circularity

### LINKED ISSUE #46582: #46218 breaks ramda types on DT
***Edit from DanielRosenwasser: #46218 #46218 #46218 #46218 #46218 #46218***

I'm working on a standalone repro, but you can see the error now by opening types/ramda/tools.d.ts and looking at lines 149 and 179:

```ts
export type Evolvable<E extends Evolver> = {
    [P in keyof E]?: Evolved<E[P]>;
};
// ......
export type Evolver<T extends Evolvable<any> = any> = {
    // if T[K] isn't evolvable, don't allow nesting for that property
    [key in keyof Partial<T>]: ((value: T[key]) => T[key]) | (T[key] extends Evolvable<any> ? Evolver<T[key]> : never);
};
```

There are now errors on constraints for `E` and `T` saying that "Type parameter 'E' has a circular constraint.".

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #46582.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Review by andrewbranch — APPROVED** [review]
Nice find 🪐

**Review by sandersn — APPROVED** [review]

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16572,7 +16572,8 @@ namespace ts {
                         if (t.flags & (TypeFlags.AnyOrUnknown | TypeFlags.InstantiableNonPrimitive | TypeFlags.Object | TypeFlags.Intersection) && t !== wildcardType && !isErrorType(t)) {
                             if (!type.declaration.nameType) {
                                 let constraint;
-                                if (isArrayType(t) || (t.flags & TypeFlags.Any) && (constraint = getConstraintOfTypeParameter(typeVariable)) && everyType(constraint, or(isArrayType, isTupleType))) {
+                                if (isArrayType(t) || t.flags & TypeFlags.Any && findResolutionCycleStartIndex(typeVariable, TypeSystemPropertyName.ImmediateBaseConstraint) < 0 &&
+                                    (constraint = getConstraintOfTypeParameter(typeVariable)) && everyType(constraint, or(isArrayType, isTupleType))) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:16576]: One day, we should find a general way to easily ask "is this true *assuming that there are no circularities?"

Today is not that day.

**Review by DanielRosenwasser — APPROVED** [review]

File: tests/baselines/reference/mappedTypeWithAny.js [hunk 2]
```diff
@@ -51,10 +51,21 @@ declare function stringifyArray<T extends readonly any[]>(arr: T): { -readonly [
...
+
+// Repro from #46582
+
+export type Evolvable<E extends Evolver> = {
+  [P in keyof E]: never;
+};
+export type Evolver<T extends Evolvable<any> = any> = {
+  [key in keyof Partial<T>]: never;
+};
+
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/mappedTypeWithAny.js:64]: Can you remove the export here, or make all of the types exported? This turned off `.d.ts` emit for the rest of the file.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: tests/baselines/reference/mappedTypeWithAny.js (hunk 2 above)
Comment (ahejlsberg) [reply tests/baselines/reference/mappedTypeWithAny.js]: Done!

**Comment by ahejlsberg** [comment]
DT tests look good, error in `ramda` is now gone.

**Pull request merged**